### ISSUE TITLE
Site Settings not updated when running on multiple instances

### ISSUE THREAD

**Issue by MarkvDijk** [episerver/Foundation#895]
When running the site on multiple instances, Site Settings are not replicated to the memory of other instances.

When updating a property in the Site Settings and publishing the content, the update is stored in the database. However the SettingsService stores a copy of the settings in memory  `public ConcurrentDictionary<string, Dictionary<Type, object>> SiteSettings`
This is done in the `PublishContent` event using `UpdateSettings(...)`. This event is only triggered on the instance you are connected to. The other instances don't trigger this event, therefore not updating the SiteSettings dictionary and returning old settings.

Issue can be replicated by using 2 browsers. Check the ARRAffinity cookie for the instance Id you are connect to. They must be different for both browsers. In both browsers open the same page that displays one of the site settings. Update the site setting in browser 1 and refresh the page. the updated setting is visible. Refresh the page in browser 2, it still displays the old setting. 

Browser 2 won't display the new setting unless you publish it from that browser as well (causing the dictionary to update)

**Comment by Hinneman** [comment]
We have exactly the same problem using Site Settings in DXP Production.

Any idea on how to solve this?

**Comment by lunchin** [comment]
Can you please check the diff of the pull requests if this fixes your issues before I merge.

**Comment by lunchin** [comment]
Hold merge until confirmation fixes issue

**Comment by Hinneman** [comment]
I will implement this change and get back to you when verified.

**Comment by Hinneman** [comment]
We have now implemented the changes, deployed them to preprod, scaled up environment and verified with two browsers with different ARRAffinity cookies. 

A change to the web site settings is still only shown in the browser where the change was published not the other one.

**Comment by jonascarlbaum** [comment]
My fix was mostly solved by updating `GetSiteSettings`-method like this.

```
public SettingsBase GetSiteSettings<T>(Guid? siteId = null) where T : SettingsBase
{
    if (!siteId.HasValue)
    {
        siteId = ResolveSiteId();
        if (siteId == Guid.Empty)
        {
            return default;
        }
    }
    try
    {
        if (PageEditing.PageIsInEditMode)
        {
            return GetEditModeSettings<T>(siteId.Value.ToString() + "-common-draft");
        }
        else
        {
            return GetViewModeSettings<T>(siteId.Value.ToString());
        }
    }
    catch (KeyNotFoundException keyNotFoundException)
    {
        _log.Error($"[Settings] {keyNotFoundException.Message}", exception: keyNotFoundException);
    }
    catch (ArgumentNullException argumentNullException)
    {
        _log.Error($"[Settings] {argumentNullException.Message}", exception: argumentNullException);
    }

    return default;
}

private T GetEditModeSettings<T>(string cacheKey) where T : SettingsBase
{
    return _synchronizedObjectInstanceCache.ReadThrough<T>($"{typeof(T).Name}:{cacheKey}",
        () =>
        {
            UpdateSettings();

            if (SiteSettings.TryGetValue(cacheKey, out var siteSettings))
            {
                if (siteSettings.TryGetValue(typeof(T), out var setting))
                {
                    return (T)setting;
                }
            }

            return default;
        },
        (result) => new CacheEvictionPolicy(TimeSpan.FromMinutes(1), CacheTimeoutType.Absolute, null, new[] { Settings.MASTERKEY }),
        ReadStrategy.Wait
    );
}

private T GetViewModeSettings<T>(string cacheKey) where T : SettingsBase
{
    return _synchronizedObjectInstanceCache.ReadThrough($"{typeof(T).Name}:{cacheKey}",
        () =>
        {
            UpdateSettings();

            if (SiteSettings.TryGetValue(cacheKey, out var siteSettings) && siteSettings.TryGetValue(typeof(T), out var setting))
            {
                return (T)setting;
            }

            return default;
        },
        (result) => new CacheEvictionPolicy(TimeSpan.FromMinutes(10), CacheTimeoutType.Absolute, null, new[] { Settings.MASTERKEY }),
        ReadStrategy.Wait
    );
}
```

and on ContentEvents when changing a setting
```
_synchronizedObjectInstanceCache.Remove(Settings.MASTERKEY);
```

Works well for us.

**Comment by optidada** [comment]
Since some time (1-2 years?) this issue is generating a lot of high priority support cases and headache with Optimizely Support. It seems like there is a proper fix/workaround since some time back, but can we have it merged as well? Thank you. @daniel-isaacs @lunchin

**Comment by jonascarlbaum** [comment]
@optidada my [code](https://github.com/episerver/Foundation/issues/895#issuecomment-2173627151) work in DXP.

I didn’t do a PR, since we had customized changes and I didn’t have the time to clone and fix. I guess anyone can do it, from my comment, and do a PR!?

**Comment by optidada** [comment]
Thanks @jonascarlbaum. You're absolutely right. Just to be clear - I wasn't necessarily asking you for a PR but anyone really that feel that it's beneficial for the project. Every solution making use of this SettingsService on Azure (with multiple instances) seems to hit this bug at some point :)

**Comment by jonascarlbaum** [comment]
> Thanks @jonascarlbaum. You're absolutely right. Just to be clear - I wasn't necessarily asking you for a PR but anyone really that feel that it's beneficial for the project. Every solution making use of this SettingsService on Azure (with multiple instances) seems to hit this bug at some point :)

Yep, on every update of a setting only the instance the editor was on will be ok, all other instances would return old settings.

As a workaround, before we implemented the fix, was restarting the site in PaaS-portal, after someone changed a setting, which wasn’t a very nice…

So, it would very much need an update, if this is a project still being cloned and referenced etc., which I guess is the case.

But maybe this project isn’t a high priority!? 🤔🤷‍♂️

**Comment by jonascarlbaum** [comment]
> @jonascarlbaum we have done the fix suggested however we are stuck with how to handle or pass this Settings.MASTERKEY `_synchronizedObjectInstanceCache.Remove(Settings.MASTERKEY);` Appreciate your valuable suggestion on this, I am getting compile time error if used code as is The type or namespace name 'MASTERKEY' does not exist in the namespace 'Foundation.Infrastructure.Cms.Settings' (are you missing an assembly reference?) Thank you so much in advance

@avinashmpawar I only added this in the bottom of `SettingsService.cs`...

```csharp
public static class Settings
{
    public static string MASTERKEY = "MASTERKEY";
}
```

**Comment by lunchin** [comment]
If someone check [PR ](https://github.com/episerver/Foundation/pull/971)to see it works for them, @daniel-isaacs  can merge this.

**Comment by jonascarlbaum** [comment]
Great @lunchin!
Haven’t checked this out, but looking at the code you kind of did as I did, in terms of updating/invalidating on other instances, but avoided using cache as invalidation trigger, you raised events for other instances to act on. Great, I did the faster implementation (since I was used to work with cache-invalidation), to achieve this. Your solution looks more fit for this use-case. 👍

**Comment by avinashmpawar** [comment]
Thank you @lunchin & @jonascarlbaum for your prompt help. Much appreciated. We do have language based translated settings, we are facing some minor issue, it's pulling other language header setting
